---
title: "Reproducing Results from Gerard and Stephens (2017)"
bibliography: readme_bib.bib
output: github_document
---

# Introduction

This repository contains code to reproduce the empirical evaluations
of @gerard2017empirical. The new methods can be found in the
[vicar](https://github.com/dcgerard/vicar) package.

```{r, echo = FALSE, results = "hide", message=FALSE, warning=FALSE}
liblist <- list("tidyverse", "dplyr", "stringr", "ggplot2", "reshape2", "pROC",
                "ruv", "cate", "gridExtra", "snow", "devtools", "sva",
                "limma", "vicar", "Rmpi", "seqgendiff", "ashr")
lapply(liblist, require, character.only = TRUE)
```

If you are having trouble reproducing these results, it might be that you need to update some of your R packages. These are the versions that I used:

```{r}
sessionInfo()
```

As you can see above, I've also only tried this out on Ubuntu.

If you find a bug, please create an
[issue](https://github.com/dcgerard/ruvb_sims/issues).

# Instructions

To reproduce the results of @gerard2017empirical, you need to (1)
install the appropriate R packages, (2) obtain the appropriate data,
(3) run `make` and (4) get some coffee while you wait.

## Install R Packages

To install the needed R packages, run the following in R

```{r, eval=FALSE}
install.packages(c("tidyverse", "stringr", "reshape2", "pROC",
                   "ruv", "cate", "gridExtra", "snow", "devtools", 
                   "Rmpi", "ashr"))
source("https://bioconductor.org/biocLite.R")
biocLite(c("sva", "limma"))
devtools::install_github("dcgerard/seqgendiff")
devtools::install_github("dcgerard/vicar")
```


## Get Data

Place the following files in the Data folder:

1. [GTEx_Data_V6_Annotations_SampleAttributesDS.txt](http://www.gtexportal.org/home/datasets#filesetFilesDiv21)
2. [GTEx_Analysis_v6p_RNA-seq_RNA-SeQCv1.1.8_gene_reads.gct.gz](http://www.gtexportal.org/home/datasets#filesetFilesDiv11)
3. [gencode.v19.genes.V6p_model.patched_contigs.gtf](http://www.gtexportal.org/home/datasets#filesetFilesDiv14)
4. [GTEx_Data_V6_Annotations_SubjectPhenotypesDS.txt](http://www.gtexportal.org/home/datasets#datasetDiv2)
5. [HK_genes.txt](http://www.tau.ac.il/~elieis/HKG/HK_genes.txt)
6. gene2ensembl.gz at ftp://ftp.ncbi.nih.gov/gene/DATA/

1 through 4 of the above are only available if you are a registered
user of the GTEx Portal. I don't think I'm allowed to release these
data.

## Run Make

To reproduce all of the results in @gerard2017empirical, simply run
`make` from the terminal.

If you want to reproduce just the results from Section 5.1, run
``` bash
make sims
```

If you want to reproduce just the results from Section 5.2, run
``` bash
make gtex_analysis
```

If you want to reproduce the figure in the introduction, run
``` bash
make one_data
```

## Get Coffee

All of these runs (except the last one) should take a very long time
(a day to a couple of days). You should get some coffee. Here is a list of some of my favorite places:

* Chicago
    - [Sawada Coffee](https://www.yelp.com/biz/sawada-coffee-chicago)
    - [Plein Air Cafe](https://www.yelp.com/biz/plein-air-cafe-and-eatery-chicago-2)
* Seattle
    - [Bauhaus Ballard](https://www.yelp.com/biz/bauhaus-ballard-seattle)
    - [Cafe Solstice](https://www.yelp.com/biz/cafe-solstice-seattle)
* Columbus
    - [Yeah, Me Too](https://www.yelp.com/biz/yeah-me-too-columbus)
    - [Stauf's Coffee Roasters](https://www.yelp.com/biz/staufs-coffee-roasters-columbus-2)
    - [Caffe Apropos](https://www.yelp.com/biz/caff%C3%A9-apropos-columbus-2)

# References
